invoices/settlement_interceptor.go

```diff
@@ -0,0 +1,224 @@
+		err := cb(clientRequest)
+		if err != nil {
+			err = fmt.Errorf("client callback failed: %w", err)
+			log.Error(err)
+			session.ClientErrChannel <- err
+		}
+	}(s.callback)
+
+	// Return the session to the caller so that they can block until the
+	// resolution is received.
+	return fn.Some(session)
+}
+
+// Modify changes parts of the HTLC based on the client's response.
+func (s *HtlcModificationInterceptor) Modify(htlc CircuitKey,
+	amountPaid lnwire.MilliSatoshi) error {
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Retrieve the intercept session for the invoice payment hash.
+	session, ok := s.activeSessions[htlc]
+	if !ok {
+		return fmt.Errorf("invoice intercept session not found "+
+			"(circuit_key=%s)", htlc)
+	}
+
+	// Send the resolution to the session resolution channel.
+	resolution := HtlcModifyResponse{
+		AmountPaid: amountPaid,
+	}
+	sendSuccessful := fn.SendOrQuit(
+		session.ClientResponseChannel, resolution, session.Quit,
+	)
+	if !sendSuccessful {
+		return fmt.Errorf("failed to send modification to client")
+	}
+
+	return nil
+}
+
+// SetClientCallback sets the client callback function that will be called when
+// an invoice is intercepted.
+func (s *HtlcModificationInterceptor) SetClientCallback(
+	callback HtlcModifyCallback) {
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.callback = callback
+}
+
+// QuitSession closes the quit channel for the session associated with the
+// given invoice. This signals to the client that the session has ended.
+func (s *HtlcModificationInterceptor) QuitSession(
+	session InterceptSession) error {
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Retrieve the intercept session and delete it from the local cache.
+	sessionKey := session.ExitHtlcCircuitKey
+	activeSession, ok := s.activeSessions[sessionKey]
+	if !ok {
+		// If the session is not found, no further action is necessary.
+		return nil
+	}
+
+	// Send to the quit channel to signal the client that the session has
+	// ended.
+	close(activeSession.Quit)
+	delete(s.activeSessions, sessionKey)
+
+	return nil
+}
+
+// QuitActiveSessions quits all active sessions by sending on each session quit
+// channel.
+func (s *HtlcModificationInterceptor) QuitActiveSessions() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for key := range s.activeSessions {
+		session := s.activeSessions[key]
+		close(session.Quit)
+		delete(s.activeSessions, key)
+	}
+
+	return nil
+}
+
+// Start starts the service.
+func (s *HtlcModificationInterceptor) Start() error {
+	return nil
+}
+
+// Stop stops the service.
+func (s *HtlcModificationInterceptor) Stop() error {
+	// If the service is stopping, we will quit all active sessions.
+	return s.QuitActiveSessions()
+}
+
+// Ensure that HtlcModificationInterceptor implements the HtlcAcceptor and
+// HtlcModifier interfaces.
+var _ HtlcModifier = (*HtlcModificationInterceptor)(nil)
+
+// MockHtlcModifier is a mock implementation of the HtlcModifier interface.
+type MockHtlcModifier struct {
+}
```

Comment:
Unfortunately in Go we can't import test code from one package into the other. And because we need the mock in multiple packages, it needs to be defined in non-test code.